SharpToken/Lib/Encoding.cs

```diff
@@ -57,61 +73,55 @@ public static GptEncoding GetEncodingForModel(string modelName)
-                escapedTokens.Add(Regex.Escape(token));
-            }
-
-            var inner = string.Join("|", escapedTokens);
-            return $"({inner})";
+            var (_, count) = EncodeCore(lineToEncode, allowedSpecial, disallowedSpecial, countOnly: true);
+            return count;
         }
 
-        public List<int> Encode(string lineToEncode,
+        public List<int> Encode(
+#if NET8_0_OR_GREATER
+            ReadOnlySpan<char> lineToEncode,
+#else
+            string lineToEncode,
+#endif
             ISet<string> allowedSpecial = null,
-            ISet<string> disallowedSpecial = null)
+            ISet<string> disallowedSpecial = null
+        )
         {
-            var specialTokensSet = new HashSet<string>(_specialTokenMappings.Keys);
-
-            if (allowedSpecial == null)
-            {
-                allowedSpecial = new HashSet<string>();
-            }
-
-            if (disallowedSpecial == null)
-            {
-                disallowedSpecial = new HashSet<string> { "all" };
-            }
-
-            if (disallowedSpecial.Contains("all"))
-            {
-                disallowedSpecial = new HashSet<string>(specialTokensSet);
-                disallowedSpecial.ExceptWith(allowedSpecial);
-            }
-
-            if (allowedSpecial.Contains("all"))
-            {
-                allowedSpecial = specialTokensSet;
-            }
-
-            if (disallowedSpecial.Count > 0)
-            {
-                var disallowedSpecialFrozen = new HashSet<string>(disallowedSpecial);
-                var regexPattern = SpecialTokenRegex(disallowedSpecialFrozen);
-                var match = Regex.Match(lineToEncode, regexPattern);
-                if (match.Success)
-                {
-                    throw new ArgumentException($"Disallowed special token found: {match.Value}");
-                }
-            }
+            var (tokens, _) = EncodeCore(lineToEncode, allowedSpecial, disallowedSpecial, countOnly: false);
+            return tokens;
+        }
 
-            var encodedLine = _bytePairEncodingCoreProcessor.EncodeNative(lineToEncode, allowedSpecial);
-            return encodedLine.Item1;
+#if NET8_0_OR_GREATER
+        // keep this overload because it was part of previous public API.
+        // It could be removed if desired.
+        public List<int> Encode(
+            string lineToEncode,
+            ISet<string> allowedSpecial = null,
+            ISet<string> disallowedSpecial = null
+        )
+        {
+            var (tokens, _) = EncodeCore(lineToEncode.AsSpan(), allowedSpecial, disallowedSpecial, countOnly: false);
+            return tokens;
         }
+#endif
 
+        // keep this overload because it was part of previous public API:
         public string Decode(List<int> inputTokensToDecode)
+        {
+            return Decode((IEnumerable<int>) inputTokensToDecode);
+        }
+
+        public string Decode(IEnumerable<int> inputTokensToDecode)
```

Comment:
Adding an `IEnumerable<int>` overload is more convenient then requiring a list allocation for the `List<int>` overload.
Especially when decoding iterates only once over the sequence.